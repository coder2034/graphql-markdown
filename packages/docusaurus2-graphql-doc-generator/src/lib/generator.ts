import { promises as fs } from "fs";
import path from "path";

import prettier from "prettier";

import { generateMarkdownFromSchema } from "@edno/graphql-markdown";

import { PluginOptions, Sidebar } from "../types";

const SIDEBAR_FILE = "sidebar-schema.js";

const prettifyJSON = (content: string): string => {
  return prettier.format(content, { parser: "json" });
};

const prettifyMarkdown = (content: string): string => {
  return prettier.format(content, { parser: "markdown" });
};

const renderHomepage = async (
  homepageLocation: string,
  options: PluginOptions
): Promise<string> => {
  const homepage = await fs.readFile(homepageLocation, "utf8");

  const filePath = path.join(options.rootPath, path.basename(homepageLocation));
  await fs.writeFile(filePath, prettifyMarkdown(homepage), "utf8");

  return path.relative("./", filePath);
};

const renderSidebar = async ({
  rootPath,
  baseURL,
}: PluginOptions): Promise<string> => {
  const sidebar: Record<string, Sidebar> = {
    schemaSidebar: [
      {
        dirName: baseURL,
        type: "autogenerated",
      },
    ],
  };

  const filePath = path.join(rootPath, SIDEBAR_FILE);
  await fs.writeFile(filePath, prettifyJSON(JSON.stringify(sidebar)), "utf8");

  return path.relative("./", filePath);
};

export const generateDocFromSchema = async (
  schemaLocation: string,
  options: PluginOptions
): Promise<{ pages: number; sidebar: string }> => {
  const pages = await generateMarkdownFromSchema({ saveToFiles: true });

  // eslint-disable-next-line no-magic-numbers
  if (typeof pages === "undefined" || pages.length === 0) {
    throw new Error("Parsing error, no pages generated");
  }

  if (typeof options === "undefined") {
    throw new Error("Parsing error, no pages generated");
  }

  await renderHomepage(options.homepage, options);
  const sidebarPath = await renderSidebar(options);

  return { pages: pages.length, sidebar: sidebarPath };
};
