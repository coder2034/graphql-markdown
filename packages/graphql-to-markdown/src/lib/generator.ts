import * as chalk from "chalk";

import {
  Renderer,
  Printer,
  getSchemaMap,
  loadSchema,
  GraphQLFileLoader,
  UrlLoader,
  JsonFileLoader,
  checkSchemaChanges,
  saveSchemaHash,
  saveSchemaFile,
  round,
} from ".";

const time = process.hrtime();

export async function generateDocFromSchema(
  baseURL: string,
  schemaLocation: string,
  outputDir: string,
  linkRoot: string,
  diffMethod: string,
  tmpDir: string,
): Promise<void> {
  const schema = await loadSchema(schemaLocation, {
    loaders: [new GraphQLFileLoader(), new UrlLoader(), new JsonFileLoader()],
  });

  const renderer: Renderer = new Renderer(
    new Printer(schema, baseURL, linkRoot),
    outputDir,
    baseURL,
  );

  const hasChanges = await checkSchemaChanges(schema, tmpDir, diffMethod);

  if (hasChanges) {
    const rootTypes: any = getSchemaMap(schema);

    const generatedPages = await Promise.all(
      Object.keys(rootTypes).map((typeName) =>
        renderer.renderRootTypes(typeName, rootTypes[typeName]),
      ),
    );

    const pages = generatedPages.reduce(
      (r: any, i: any) => [].concat(r, i),
      [],
    );

    const sidebarPath = renderer.renderSidebar(pages);

    const [sec, msec] = process.hrtime(time);
    const duration = round(sec + msec / 1000000000, 3);
    console.info(
      chalk.green(
        `Documentation successfully generated in "${outputDir}" with base URL "${baseURL}".`,
      ),
    );
    console.log(
      chalk.blue(
        `${pages.length} pages generated in ${duration}s from schema "${schemaLocation}".`,
      ),
    );
    console.info(
      chalk.blue.bold(
        `Remember to update your Docusaurus site's sidebars with "${sidebarPath}".`,
      ),
    );
  } else {
    console.info(
      chalk.blue(`No changes detected in schema "${schemaLocation}".`),
    );
  }
  // create references for checkSchemaChanges
  await saveSchemaHash(schema, tmpDir);
  await saveSchemaFile(schema, tmpDir);
}
