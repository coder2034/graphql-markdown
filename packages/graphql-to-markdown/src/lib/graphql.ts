import {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  GraphQLFieldMap,
  GraphQLField,
  GraphQLSchema,
  GraphQLNamedType,
} from "graphql";

import { Maybe } from "graphql/jsutils/Maybe";
import { TypeMap } from "graphql/type/schema";

import { hasOwnMethod, hasOwnProperty } from ".";

export {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  isDirective as isDirectiveType,
  getNamedType,
  isScalarType,
  isEnumType,
  isUnionType,
  isInterfaceType,
  isObjectType,
  isInputObjectType as isInputType,
  isNullableType,
  printSchema,
} from "graphql";
export { loadSchema } from "@graphql-tools/load";
export { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
export { UrlLoader } from "@graphql-tools/url-loader";
export { JsonFileLoader } from "@graphql-tools/json-file-loader";

export const SCHEMA_EXCLUDE_LIST_PATTERN =
  /^(?!Query$|Mutation$|Subscription$|__.+$).*$/;

export function getDefaultValue(argument: { type: unknown; defaultValue: unknown }) {
  if (isListType(argument.type)) {
    return `[${argument.defaultValue || ""}]`;
  }

  switch (argument.type) {
    case GraphQLID:
    case GraphQLInt:
      return `${argument.defaultValue || "0"}`;
    case GraphQLFloat:
      return `${argument.defaultValue || "0.0"}`;
    case GraphQLString:
    default:
      return argument.defaultValue ? `"${argument.defaultValue}"` : undefined;
  }
}

export function getFilteredTypeMap(
  typeMap: Record<string, GraphQLNamedType>,
  excludeList = SCHEMA_EXCLUDE_LIST_PATTERN,
): Maybe<TypeMap> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => excludeList.test(key))
    .reduce((res: TypeMap, key) => ((res[key] = typeMap[key]), res), {});
}

export function getIntrospectionFieldsList(queryType: Maybe<GraphQLObjectType<unknown, unknown>>): Maybe<GraphQLFieldMap<unknown, unknown>> {
  if (!queryType){
    return;
  }

  if (!hasOwnMethod(queryType, "getFields")) {
    return;
  }
  return queryType.getFields();
}

export function getFields(type: GraphQLObjectType<unknown, unknown>): Maybe<GraphQLField<any, unknown, Record<string, unknown>>[]> {
  if (!hasOwnMethod(type, "getFields")) {
    return;
  }
  const fieldMap = type.getFields();
  return Object.keys(fieldMap).map((name) => fieldMap[name]);
}

export function getTypeName(
  type: GraphQLObjectType<unknown, unknown>,
  defaultName = "",
): Maybe<string> {
  if (!type) {
    return;
  }
  return (
    (hasOwnProperty(type, "name") && type.name) ||
    (hasOwnMethod(type, "toString") && type.toString()) ||
    defaultName
  );
}

export function getTypeFromTypeMap(typeMap: TypeMap, type: any): Maybe<Record<string, unknown>> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => typeMap[key] instanceof type)
    .reduce((res: any, key) => ((res[key] = typeMap[key]), res), {});
}

export function getSchemaMap(schema: GraphQLSchema) {
  const typeMap = getFilteredTypeMap(schema.getTypeMap());
  if (!typeMap) return;
  return {
    queries: getIntrospectionFieldsList(
      schema.getQueryType && schema.getQueryType(),
    ),
    mutations: getIntrospectionFieldsList(
      schema.getMutationType && schema.getMutationType(),
    ),
    subscriptions: getIntrospectionFieldsList(
      schema.getSubscriptionType && schema.getSubscriptionType(),
    ),
    directives: schema.getDirectives(),
    objects: getTypeFromTypeMap(typeMap, GraphQLObjectType),
    unions: getTypeFromTypeMap(typeMap, GraphQLUnionType),
    interfaces: getTypeFromTypeMap(typeMap, GraphQLInterfaceType),
    enums: getTypeFromTypeMap(typeMap, GraphQLEnumType),
    inputs: getTypeFromTypeMap(typeMap, GraphQLInputObjectType),
    scalars: getTypeFromTypeMap(typeMap, GraphQLScalarType),
  };
}

export function isParametrizedField(field: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(field, "args") && field.args.length > 0;
}

export function isOperation(query: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(query, "type");
}
