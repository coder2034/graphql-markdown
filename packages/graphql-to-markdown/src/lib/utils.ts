import slugify from "slugify";
import { kebabCase } from "lodash";

export { kebabCase, startCase, round } from "lodash";

export function toSlug(input: string): string {
  return slugify(kebabCase(input));
}

// from https://fettblog.eu/typescript-hasownproperty/
export function hasOwnProperty<X extends unknown, Y extends PropertyKey>(obj: X, prop: Y): obj is X & Record<Y, unknown> {
  return typeof obj === "object" && Object.prototype.hasOwnProperty.call(obj, prop);
}

export function hasOwnMethod<X extends unknown, Y extends PropertyKey>(obj: X, prop: Y): obj is X & Record<Y, unknown> {
  return hasOwnProperty(obj, prop) && typeof obj[prop] === "function";
}
