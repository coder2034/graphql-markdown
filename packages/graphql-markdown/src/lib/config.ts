import {
  GraphQLExtensionDeclaration,
  GraphQLProjectConfig,
  loadConfig,
} from "graphql-config";
import { DocumentNode } from "graphql";

import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { JsonFileLoader } from "@graphql-tools/json-file-loader";
import { UrlLoader } from "@graphql-tools/url-loader";

import { ExtensionAPI, LoadConfigOptions, Maybe } from "..";

const EXTENSION_NAME = "graphql-markdown" as const;

export const OPTION = {
  LAYOUTS: "layouts",
  MDX: "mdx",
  OUTPUT: "output",
};

export type ConfigurationOptions = {
  excludes?: readonly string[];
  layouts?: string;
  mdx?: boolean;
  output?: string;
};

const defaultOptions: ConfigurationOptions = {
  excludes: [] as const,
  [OPTION.LAYOUTS]: "./layouts" as const,
  [OPTION.OUTPUT]: false as const,
  output: "./output" as const,
} as const;

const setFileLoaderExtension: GraphQLExtensionDeclaration = (
  api: ExtensionAPI
) => {
  [new GraphQLFileLoader(), new JsonFileLoader(), new UrlLoader()].forEach(
    (loader) => {
      return api.loaders.schema.register(loader);
    }
  );
  return { name: EXTENSION_NAME };
};

const loadConfiguration = async (
  options?: LoadConfigOptions
): Promise<Maybe<GraphQLProjectConfig>> => {
  const config = await loadConfig({
    ...options,
    extensions: [setFileLoaderExtension],
  });
  return config?.getDefault();
};

export class Configuration {
  static project: Maybe<GraphQLProjectConfig>;
  static extension: ConfigurationOptions;

  static load = async (options?: LoadConfigOptions): Promise<void> => {
    const configuration = await loadConfiguration(options);
    Configuration.project = configuration;
    Configuration.extension = {
      ...defaultOptions,
      ...Configuration.project?.extension(EXTENSION_NAME),
    };
  };

  static get = (name: string): string => {
    return Configuration.extension[name];
  };

  static set = (name: string, value: string): void => {
    Configuration.extension[name] = value;
  };

  // eslint-disable-next-line require-await
  static schema = async (): Promise<Maybe<DocumentNode>> => {
    return Configuration.project!.getSchema("DocumentNode");
  };
}
