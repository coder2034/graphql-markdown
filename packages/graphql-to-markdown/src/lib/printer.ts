import { GraphQLArgument, GraphQLEnumValue, GraphQLField, GraphQLSchema, GraphQLType } from "graphql";
import * as path from "path";
import {
  isEnumType,
  isUnionType,
  isObjectType,
  isScalarType,
  isOperation,
  getDefaultValue,
  getTypeName,
  getFields,
  isDirectiveType,
  isParametrized,
  isInterfaceType,
  getNamedType,
  isInputType,
  isListType,
  toSlug,
  hasOwnProperty,
  hasOwnMethod,
  prettifyMarkdown,
} from ".";

const HEADER_SECTION_LEVEL = "###";
const HEADER_SECTION_SUB_LEVEL = "####";
const HEADER_SECTION_ITEM_LEVEL = "- #####";
const NO_DESCRIPTION_TEXT = "No description";

const TAG = `
export const Tag = ({children, color}) => (
  <span
    style={{
      backgroundColor: color,
      borderRadius: '2px',
      color: '#fff',
      padding: '0.2rem',
    }}>
    {children}
  </span>
);`;

export class Printer {
  constructor(readonly schema: GraphQLSchema, readonly baseURL: string, readonly linkRoot = "/") {
  }

  toLink(type: any, name: string): string {
    let graphLQLNamedType: GraphQLType = getNamedType(<GraphQLType>type);
    if (isListType(type)) graphLQLNamedType = getNamedType(<GraphQLType>type.ofType);
    let category;
    switch (true) {
      case isEnumType(graphLQLNamedType):
        category = "enums";
        break;
      case isUnionType(graphLQLNamedType):
        category = "unions";
        break;
      case isInterfaceType(graphLQLNamedType):
        category = "interfaces";
        break;
      case isObjectType(graphLQLNamedType):
        category = "objects";
        break;
      case isInputType(graphLQLNamedType):
        category = "inputs";
        break;
      case isScalarType(graphLQLNamedType):
        category = "scalars";
        break;
      case isDirectiveType(graphLQLNamedType):
        category = "directives";
        break;
    }

    if (category && graphLQLNamedType) {
      return `[\`${name}\`](${path.join(
        this.linkRoot,
        this.baseURL,
        category,
        toSlug(graphLQLNamedType.toString()),
      )})`;
    } else {
      return `\`${name}\``;
    }
  }

  printSection(values: GraphQLType[] | GraphQLEnumValue[] | GraphQLArgument[], section: string, level = HEADER_SECTION_LEVEL): string {
    if (values.length > 0)
      return `${level} ${section}\n\n${this.printSectionItems(values)}\n\n`;
    return "";
  }

  printSectionItems(values: GraphQLType[] | GraphQLArgument[] | GraphQLEnumValue[], level = HEADER_SECTION_SUB_LEVEL): string {
    if (Array.isArray(values))
      return values
        .map((v: GraphQLType | GraphQLArgument | GraphQLEnumValue) => v && this.printSectionItem(v, level))
        .join(`\n\n`);
    return "";
  }

  printSectionItem(type: GraphQLType | GraphQLArgument | GraphQLEnumValue, level = HEADER_SECTION_SUB_LEVEL): string {
    if (!type) {
      return "";
    }

    let section = `${level} ${this.toLink(type, getTypeName(type))} ${
      hasOwnProperty(type, "type")
        ? `(${this.toLink(<GraphQLType>type.type, getTypeName(<GraphQLType>type.type))})`
        : ""
    }\n\n${this.printDescription(type, "")}\n`;
    if (isParametrized(type)) {
      section += this.printSectionItems(type.args, HEADER_SECTION_ITEM_LEVEL);
    }
    return section;
  }

  printCodeEnum(type: GraphQLType): string {
    let code = "";
    if (isEnumType(type)) {
      code += `enum ${getTypeName(type)} {\n`;
      code += type
        .getValues()
        .map((value) => `  ${getTypeName(value)}`)
        .join("\n");
      code += `\n}`;
    }
    return code;
  }

  printCodeUnion(type: GraphQLType): string {
    let code = "";
    if (isUnionType(type)) {
      code += `union ${getTypeName(type)} = `;
      code += type
        .getTypes()
        .map((v) => getTypeName(v))
        .join(" | ");
    }
    return code;
  }

  printCodeScalar(type: GraphQLType): string {
    return `scalar ${getTypeName(type)}`;
  }

  printCodeArguments(type: any): string {
    let code = "";
    if (isParametrized(type)) {
      code += `(\n`;
      code += type.args.reduce(
        (r: never, v: GraphQLArgument ) => {
          const defaultValue = getDefaultValue(v);
          return `${r}  ${getTypeName(<GraphQLType><unknown>v)}: ${v.type.toString() ?? ""}${
            defaultValue ? ` = ${defaultValue}` : ""
          }\n`;
        },
        "",
      );
      code += `)`;
    }
    return code;
  }

  printCodeField<TSource, TContext, TArgs>(type: GraphQLField<TSource, TContext, TArgs>): string {
    let code = `${getTypeName(<GraphQLType><unknown>type)}`;
    code += this.printCodeArguments(type);
    if (hasOwnProperty(type, "type")) code += `: ${getTypeName(<GraphQLType>type.type)}\n`;
    return code;
  }

  printCodeDirective(type: GraphQLType): string {
    let code = `directive @${getTypeName(type)}`;
    code += this.printCodeArguments(type);
    return code;
  }

  printCodeType(type: GraphQLType): string {
    let code = `${isInterfaceType(type) ? "interface" : "type"} ${getTypeName(
      type,
    )}`;
    code += `${
      hasOwnMethod(type, "getInterfaces") && type.getInterfaces().length > 0
        ? ` implements ${type
            .getInterfaces()
            .map((v: GraphQLType) => getTypeName(v))
            .join(", ")}`
        : ""
    }`;
    code += ` {\n`;
    code += getFields(type)
      ?.map((v: any) => `  ${this.printCodeField(v)}`)
      ?.join("");
    code += `}`;

    return code;
  }

  printHeader(id: string, title: string): string {
    return `---\nid: ${id}\ntitle: ${title}\n---\n`;
  }

  printDeprecation(type: GraphQLType | GraphQLArgument | GraphQLEnumValue): string {
    if (hasOwnProperty(type, "isDeprecated") && type.isDeprecated) {
      const reason = hasOwnProperty(type, "deprecationReason") && ` ${type.deprecationReason}` || "";
      return `<sub><sup><Tag color="#ffba00">DEPRECATED</Tag>${reason}</sup></sub>\n\n`;
    }
    return "";
  }

  printDescription(type: GraphQLType | GraphQLArgument | GraphQLEnumValue, noText = NO_DESCRIPTION_TEXT): string {
    let description = "";

    description = `${this.printDeprecation(type)}${
      (hasOwnProperty(type, "description") && type.description) || noText
    }`;

    return description;
  }

  printCode(type: GraphQLType): string {
    let code = "\n```graphql\n";
    switch (true) {
      case isEnumType(type):
        code += this.printCodeEnum(type);
        break;
      case isUnionType(type):
        code += this.printCodeUnion(type);
        break;
      case isInterfaceType(type):
      case isObjectType(type):
      case isInputType(type):
        code += this.printCodeType(type);
        break;
      case isScalarType(type):
        code += this.printCodeScalar(type);
        break;
      case isDirectiveType(type):
        code += this.printCodeDirective(type);
        break;
      case isOperation(type):
        code += this.printCodeField(<any>type);
        break;
      default:
        code += `"${getTypeName(type)}" not supported`;
    }
    code += "\n```\n";
    return code;
  }

  printType(name: string, type: GraphQLType): string {
    if (!type) {
      return "";
    }

    const header = this.printHeader(name, getTypeName(type));
    const description = this.printDescription(type);
    const code = this.printCode(type);

    let metadata = "";
    if (isEnumType(type)) {
      metadata = this.printSection(type.getValues(), "Values");
    }

    if (isUnionType(type)) {
      metadata = this.printSection(type.getTypes(), "Possible types");
    }

    if (isObjectType(type) || isInterfaceType(type) || isInputType(type)) {
      metadata = this.printSection(<GraphQLType[]>getFields(type), "Fields");
    }

    if (
      (isObjectType(type) || isInterfaceType(type)) &&
      hasOwnMethod(type, "getInterfaces")
    ) {
      metadata += this.printSection(type.getInterfaces(), "Interfaces");
    }

    if (isOperation(type)) {
      metadata = this.printSection(type.args, "Arguments");
      const queryType = getTypeName(type.type).replace(/[![\]]*/g, "");
      metadata += this.printSection([<GraphQLType>this.schema.getType(queryType)], "Type");
    }

    if (isDirectiveType(type)) {
      metadata = this.printSection(type.args, "Arguments");
    }

    return prettifyMarkdown(
      `${header}\n\n${TAG}\n\n${description}\n\n${code}\n\n${metadata}\n\n`,
    );
  }
}
