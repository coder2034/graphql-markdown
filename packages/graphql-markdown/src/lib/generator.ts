import { Configuration, ConfigurationOptions } from "./config";
import { LoadConfigOptions, ParsedNode, Result } from "..";
import { renderNode, saveMarkdownFile } from "./render";
import { parseSchema } from "./parser";

const loadNodes = async (
  options?: LoadConfigOptions
): Promise<ParsedNode[]> => {
  await Configuration.load(options);

  const schema = await Configuration.schema();

  if (typeof schema === "undefined") {
    throw new Error("An error occurred while loading the schema");
  }

  return parseSchema(schema);
};

type Options = {
  saveToFiles?: boolean;
  loaderOptions?: LoadConfigOptions;
  generatorOptions?: ConfigurationOptions;
};

export const generateMarkdownFromSchema = async (
  options?: Options
): Promise<Result[]> => {
  const nodes = await loadNodes(options?.loaderOptions);

  const processor =
    options?.saveToFiles === true ? saveMarkdownFile : renderNode;

  const results = await Promise.all(
    nodes.map(async (node: ParsedNode): Promise<Result> => {
      const result = await processor(node, options?.generatorOptions);
      return {
        name: node.name,
        type: node.type,
        /* eslint-disable prettier/prettier */
        ...(options?.saveToFiles === true
          ? { filepath: result }
          : { markdown: result }),
        /* eslint-enable */
      };
    })
  );

  return results;
};
