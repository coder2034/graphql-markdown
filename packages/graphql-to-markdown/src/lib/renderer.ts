import * as path from "path";
import * as fs from "fs-extra";
import { pull } from "lodash";

import { toSlug, startCase, hasOwnProperty } from "./utils";
import { prettifyJavascript } from "./prettier";
import { Printer } from "./printer";

const SIDEBAR = "sidebar-schema.js";
const HOMEPAGE_ID = "schema";

export class Renderer {
  constructor(readonly printer: Printer, readonly outputDir: string, readonly baseURL: string) {
    this.emptyOutputDir();
  }

  emptyOutputDir() {
    fs.emptyDirSync(this.outputDir);
  }

  async renderRootTypes(name: string, type: any) {
    let pages: ({ category: string; slug: string } | undefined)[] = [];
    if (type) {
      const slug = toSlug(name);
      const dirPath = path.join(this.outputDir, slug);
      if (Array.isArray(type)) {
        type = type.reduce(function (r, o) {
          if (o && o.name) r[o.name] = o;
          return r;
        }, {});
      }

      fs.ensureDir(dirPath);

      await Promise.all(
        Object.keys(type).map((name: string) =>
          this.renderTypeEntities(dirPath, name, type[name]),
        ),
      ).then((p) => {
        pages = [...p];
      });
    }
    return pages;
  }

  async renderTypeEntities(dirPath: string, name: string, type: any) {
    if (type) {
      const fileName = toSlug(name);
      const filePath = path.join(dirPath, `${fileName}.mdx`);
      const content = this.printer.printType(fileName, type);
      await fs.outputFile(filePath, content, "utf8");
      const page: any = path
        .relative(this.outputDir, filePath)
        .match(
          /(?<category>[A-z][A-z0-9-]*)\/(?<pageId>[A-z][A-z0-9-]*).mdx?$/,
        );
      const slug = path.join(page.groups.category, page.groups.pageId);
      return { category: startCase(page.groups.category), slug: slug };
    }
  }

  async renderSidebar(pages: any[]) {
    const filePath = path.join(this.outputDir, SIDEBAR);
    const content = prettifyJavascript(`module.exports = {
          schemaSidebar:
          ${JSON.stringify(this.generateSidebar(pages))}
        };`);
    fs.outputFileSync(filePath, content, "utf8");
    return path.relative("./", filePath);
  }

  generateSidebar(pages: any[]) {
    let graphqlSidebar = [
      { type: "doc", id: path.join(this.baseURL, HOMEPAGE_ID) },
    ];
    pages.map((page: { category: any; slug: any }) => {
      const category: any = graphqlSidebar.find(
        (entry: any) => "label" in entry && page.category == entry.label,
      );
      const items = hasOwnProperty(category, "items") ? category?.items : [];
      const slug = path.join(this.baseURL, page.slug);
      graphqlSidebar = [
        ...pull(graphqlSidebar, category),
        {
          type: "category",
          label: page.category,
          items: [...items, slug].sort(),
        },
      ];
    });
    return graphqlSidebar;
  }
}
