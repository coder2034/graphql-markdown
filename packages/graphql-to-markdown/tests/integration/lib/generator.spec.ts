import * as path from "path";
import * as mock from "mock-fs";
import * as dirTree from "directory-tree";

import { generateDocFromSchema } from "../../../src/lib/generator";

describe("lib", () => {
  beforeEach(() => {
    mock({
      node_modules: mock.load(path.resolve(__dirname, "../../../node_modules")),
      __data__: mock.load(path.resolve(__dirname, "../../__data__")),
      output: {},
    });
  });

  afterEach(() => {
    mock.restore();
  });

  describe("renderer", () => {
    describe("generateDocFromSchema()", () => {
      test("generates Markdown document structure from GraphQL schema", async () => {
        expect.assertions(2);

        await generateDocFromSchema(
          "graphql",
          "__data__/tweet.graphql",
          "output",
          "docs",
          "SCHEMA-DIFF",
          "tmp",
        );

        expect(dirTree("output")).toMatchSnapshot();
        expect(dirTree("tmp")).toMatchSnapshot();
      });
    });
  });
});
