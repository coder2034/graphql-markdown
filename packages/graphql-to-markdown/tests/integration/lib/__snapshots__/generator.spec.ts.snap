// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".mdx",
          "name": "deprecated.mdx",
          "path": "output/directives/deprecated.mdx",
          "size": 663,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "include.mdx",
          "path": "output/directives/include.mdx",
          "size": 458,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "skip.mdx",
          "path": "output/directives/skip.mdx",
          "size": 440,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "specified-by.mdx",
          "path": "output/directives/specified-by.mdx",
          "size": 471,
          "type": "file",
        },
      ],
      "name": "directives",
      "path": "output/directives",
      "size": 2032,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "enums",
      "path": "output/enums",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "inputs",
      "path": "output/inputs",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "interfaces",
      "path": "output/interfaces",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "objects",
      "path": "output/objects",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".mdx",
          "name": "boolean.mdx",
          "path": "output/scalars/boolean.mdx",
          "size": 316,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "date.mdx",
          "path": "output/scalars/date.mdx",
          "size": 266,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "id.mdx",
          "path": "output/scalars/id.mdx",
          "size": 578,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "int.mdx",
          "path": "output/scalars/int.mdx",
          "size": 380,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "string.mdx",
          "path": "output/scalars/string.mdx",
          "size": 439,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "url.mdx",
          "path": "output/scalars/url.mdx",
          "size": 263,
          "type": "file",
        },
      ],
      "name": "scalars",
      "path": "output/scalars",
      "size": 2242,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "unions",
      "path": "output/unions",
      "size": 0,
      "type": "directory",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 4274,
  "type": "directory",
}
`;

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 2`] = `
Object {
  "children": Array [
    Object {
      "extension": "",
      "name": ".schema",
      "path": "tmp/.schema",
      "size": 64,
      "type": "file",
    },
    Object {
      "extension": ".graphql",
      "name": "schema.graphql",
      "path": "tmp/schema.graphql",
      "size": 1173,
      "type": "file",
    },
  ],
  "name": "tmp",
  "path": "tmp",
  "size": 1237,
  "type": "directory",
}
`;
